components/scream/src/physics/shoc/shoc_functions_f90.hpp

```diff
@@ -17,358 +17,195 @@
-// interfaces, and pressure gradient the functon returns dz_zi, dz_zt,
-// and density. 
-void shoc_grid(Int nlev, SHOCGridData &d);
+  void randomize();
 
+  Int total() const { return m_total; }
+  Int totali() const { return m_totali; }
 
-//Create data structure to hold data for integ_column_stability
-struct SHOCColstabData {
-  static constexpr size_t NUM_ARRAYS   = 3; //# of arrays with values at cell centers (zt grid)
-  static constexpr size_t NUM_ARRAYS_c = 1; //# of arrays with column only dimensions
+ private:
+  void init_ptrs();
 
-  // Inputs
-  Int   shcol, nlev;
-  Real *dz_zt, *pres, *brunt;
+  // Internals
+  Int m_total, m_totali;
+  std::vector<Real**> m_ptrs, m_ptrs_i;
+  std::vector<Real> m_data;
+};
 
-  // Output
-  Real *brunt_int;
+struct SHOCGridData : public SHOCDataBase {
+  // Inputs
+  Real *zt_grid, *zi_grid, *pdel;
 
-  //functions to initialize data
-  SHOCColstabData(Int shcol_, Int nlev_);
-  SHOCColstabData(const SHOCColstabData &rhs);
-  SHOCColstabData &operator=(const SHOCColstabData &rhs);
+  // In/out
+  Real *dz_zt, *dz_zi, *rho_zt;
 
-  void init_ptrs();
+  SHOCGridData(Int shcol_, Int nlev_, Int nlevi_) :
+    SHOCDataBase(shcol_, nlev_, nlevi_, {&zt_grid, &dz_zt, &pdel, &rho_zt}, {&zi_grid, &dz_zi}) {}
 
-  // Internals
-  Int m_shcol, m_nlev, m_total, m_totalc;
-  std::vector<Real> m_data;
-  std::vector<Real> m_datac;
+  SHOCGridData(const SHOCGridData &rhs) : SHOCDataBase(rhs, {&zt_grid, &dz_zt, &pdel, &rho_zt}, {&zi_grid, &dz_zi}) {}
 
-  template <util::TransposeDirection::Enum D>
-  void transpose() {
-    SHOCColstabData d_trans(*this);
+  SHOCGridData &operator=(const SHOCGridData &rhs) { SHOCDataBase::operator=(rhs); return *this; }
+};
 
-    // Transpose on the zt grid
-    util::transpose<D>(dz_zt, d_trans.dz_zt, shcol, nlev);
-    util::transpose<D>(pres, d_trans.pres, shcol, nlev);
-    util::transpose<D>(brunt, d_trans.brunt, shcol, nlev);
-    
-    // Transpose on the column only grid
-    util::transpose<D>(brunt_int, d_trans.brunt_int, shcol, 1);
-
-    *this = std::move(d_trans);
-  }
-};//SHOCColstabData
+struct SHOCColstabData : public SHOCDataBase {
+  // Inputs
+  Real *dz_zt, *pres, *brunt;
 
-void integ_column_stability(Int nlev, SHOCColstabData &d);
+  // Output
+  Real *brunt_int;
 
-//Create data structure to hold data for calc_shoc_vertflux
-struct SHOCTkeshearData {
-  static constexpr size_t NUM_ARRAYS   = 2; //# of arrays with values at cell centers (zt grid)
-  static constexpr size_t NUM_ARRAYS_i = 2; //# of arrays with values at interface centers (zi grid)
+  SHOCColstabData(Int shcol_, Int nlev_) :
+    SHOCDataBase(shcol_, nlev_, 1, {&dz_zt, &pres, &brunt}, {&brunt_int}) {}
+  SHOCColstabData(const SHOCColstabData &rhs) : SHOCDataBase(rhs, {&dz_zt, &pres, &brunt}, {&brunt_int}) {}
+  SHOCColstabData &operator=(const SHOCColstabData &rhs) { SHOCDataBase::operator=(rhs); return *this; }
+};//SHOCColstabData
 
+struct SHOCTkeshearData : public SHOCDataBase {
```

Comment:
All of these *Data structs should inherit from SHOCDataBase. This saves tons of repeated code. I will think about how to further streamline this process, but this approach is good enough for now.